Python/optimizer_bytecodes.c

```diff
@@ -113,6 +115,23 @@ dummy_func(void) {
         sym_set_type(right, &PyLong_Type);
     }
 
+    op(_GUARD_TYPE_VERSION, (type_version/2, owner -- owner)) {
+        if (sym_matches_type_version(owner, type_version)) {
```

Comment:
Staring at this a bit harder, it seems like there are a couple of different cases that we want to distinguish:

1. the symbol's version is 0, the expected version is 0
2. the symbol's version is nonzero, the expected version is 0
3. the symbol's version is 0, the expected version is nonzero
4. the symbol's version is nonzero, the expected version is nonzero, both match
5. the symbol's version is nonzero, the expected version is nonzero, both don't match

Cases 1 and 2 should be impossible, and we can `assert(type_version)` above this line.

Case 3 is what we're correctly handling in the `else` currently.

Case 4 is what we're correctly handling in the `if` currently.

Case 5 should result in narrowing `owner` to the "bottom" type, right? If so, this should probably happen in `sym_set_type_version`, which is called on line 127.

@Fidget-Spinner or @markshannon, can you confirm that my understanding is correct here?

Reply:
Thanks for thinking through all of these. For case 5, I am unclear how/when this could happen. This would only occur if `_CHECK_TYPE_VERSION` was ommitted twice, the first time with one type version, and then the second time with a different type version as an arg, but for the same symbolic value.

I am a little fuzzy on how these versions are generated. Are they captured based on the current type version of the type when it's generating the opcodes? If so, how could it be two different values?

If there is an example/test case here that would be helpful. Also is it possible to see in the test runner the additional args passed to each op, like the `type_version`?

Reply:
> Are they captured based on the current type version of the type when it's generating the opcodes? If so, how could it be two different values?

Yup the type version is the one at the time of specialization (bytecode rewriting). So if it just so happens that for example, a re-specialization is happening, the bytecode might have out of sync versions in their cache entries, leading to the scenario Brandt said in 5.

In general, we should hit the bottom if we're trying to set a type on something that is contradictory. E.g. if we're setting a type version on something that already has a type version. That is a contradiction. You can see an example test here https://github.com/python/cpython/blob/82bc177e9b116b4e816957a58105ebbd3ac21f81/Python/optimizer_symbols.c#L484

Reply:
That makes sense. I changed it so that it would set it to bottom if they are different versions (and not add the watcher b/c it must have already been added then).

After this change, I wonder if the logic would be clearer if it was all just in this block instead of in the get/set helper functions.